### PR TITLE
units: add time-set.target

### LINKED ISSUE #8861: proposal: add time-set.target distinct from time-sync.target
Though time-wait-sync is a good step I don't think the problem it addresses is fully solved.

The issue is that historically systemd has documented `time-sync.target` with "All services where correct time is essential should be ordered after this unit", and used it to release SysV services that depend on the `$time` facility.  Through 238 the target was reached when timesyncd set the clock from its local timestamp file, but had not yet actually communicated with a reliable time source.

Internal to systemd, the only thing depending on `time-sync.target` is `systemd-tmpfiles-clean`.  This was added in 2014 with the explanation:

    That way, on systems lacking an RTC we don't false start removing aged
    files too early.

IMO this clock set event is reasonable for removing tmp files, but does not satisfy the "correct time" guarantee implied by the documentation of `time-sync.target`.  This guarantee is required for services that need to communicate with other systems that have a timestamp accuracy requirement for authentication: 15 minutes for AWS, 5 minutes for Kerberos (see links).  The timestamp file can't provide that.

I propose that a new target `time-set.target` be added as a special, and that the unit files for `systemd-timesyncd` and `systemd-tmpfiles-clean` both be updated to use this one instead of `time-sync.target`.  This leaves `time-sync.target` for SysV `$time` and for other services that really do need accurate clocks.

See: https://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html#RESTAuthenticationTimeStamp
See: https://msdn.microsoft.com/en-us/library/ms813592.aspx

### PR DESCRIPTION
**Pull request by pabigot**
time-sync.target is supposed to indicate system clock is synchronization
with a remote clock, but as used through 238 it only provided a system
clock that was updated based on a locally-maintained timestamp.  Systems
that are powered off for extended periods would not come up with
accurate time.

Retain the existing behavior using a new time-set.target leaving
time-sync.target for cases where accuracy is required.

Closes #8861

### CODE REVIEW

**Comment by keszybz** [comment]
Looks good mostly, but I think there should be a Wants=time-set.target, After=time-set.target dependencies in time-sync.target. Also needs a rebase.

**Comment by pabigot** [comment]
@keszybz There was some discussion on #8861 whether this was desired.  Has that been resolved to "yes"?  If so, I should be able to get to this this weekend.

**Comment by keszybz** [comment]
Your explanation in https://github.com/systemd/systemd/issues/8861#issuecomment-388133220 is convincing for me.

**Comment by pabigot** [comment]
@keszybz Rebased and tested in my environment.

I added the After/Wants to `time-sync.target`, but also had to update `time-wait-sync.service` to get a system that doesn't enable timesyncd but does use time-wait-sync to emit the right synchronization points.

**Comment by pabigot** [comment]
@keszybz would you please revisit the reviewed/needs-rework state of this PR?  I've rebased this once since rework to resolve conflicts from subsequent upstream modifications, and again just now to make sure it's still compatible.  I'd like to get this merged and visible to others before 240 is frozen.

(AFAICT the failures from this last push are due to infrastructure/master, as the version I pushed four weeks ago passed without any problems and several other recently updated PRs fail the same way.)

Thanks.

**Comment by pabigot** [comment]
@keszybz ping?  I'm willing to keep rebasing this, but I'd appreciate some feedback as to whether it's ever going to be merged or if further changes are required.  Thanks.

File: units/systemd-time-wait-sync.service.in [hunk 1]
```diff
@@ -22,8 +22,8 @@ ConditionCapability=CAP_SYS_TIME
 ConditionVirtualization=!container
 
 DefaultDependencies=no
-Before=time-sync.target shutdown.target
-Wants=time-sync.target
+Before=time-set.target time-sync.target shutdown.target
+Wants=time-set.target time-sync.target
```
Comment (poettering) [inline units/systemd-time-wait-sync.service.in]: hmm, why this change? this service waits for the ntp sync hence should only be order before classic time-sync.target

File: units/systemd-timesyncd.service.in [hunk 2]
```diff
@@ -14,9 +14,9 @@ ConditionCapability=CAP_SYS_TIME
 ConditionVirtualization=!container
 DefaultDependencies=no
 After=systemd-remount-fs.service systemd-sysusers.service
-Before=time-sync.target sysinit.target shutdown.target
+Before=time-set.target sysinit.target shutdown.target
 Conflicts=shutdown.target
-Wants=time-sync.target
+Wants=time-set.target
```
Comment (poettering) [inline units/systemd-timesyncd.service.in]: this should pull in both i think, after all services ordering themselves after time-sync.target should always be ordered between timesyncd too

File: units/systemd-time-wait-sync.service.in (hunk 1 above)
Comment (pabigot) [reply units/systemd-time-wait-sync.service.in]: At the time I thought it necessary to get the synchronization points.  Today it doesn't, so I've 
removed this change.

**Comment by keszybz** [comment]
bionic-amd64 fails in TEST-30-ONCLOCKCHANGE. This might be relevant, but that test has been flaky too. I'll restart it.

File: units/systemd-timesyncd.service.in (hunk 2 above)
Comment (pabigot) [reply units/systemd-timesyncd.service.in]: Done

**Comment by keszybz** [comment]
This seems low risk, but I'm not sure if we should merge it now or after the release.
